File: contests/16_05_2020/3.py
#!/usr/bin/env python
from typing import Optional, List
from copy import deepcopy as dc

from solutions.minions import Tree as TreeNode


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        ans = 0

        def preorder(t: Optional[TreeNode], lst: Optional[List[int]] = None) -> None:
            nonlocal ans

            if t:
                if lst is None: lst = []
                if lst and t.v >= max(lst): ans += 1
                lst.extend([t.v])
                preorder(t.left,  lst=dc(lst))
                preorder(t.right, lst=dc(lst))

        preorder(root)

        return ans + 1


if __name__ == '__main__':
    Solution().goodNodes(
        TreeNode(
            2,
            None,
            TreeNode(
                4,
                TreeNode(10),
                TreeNode(
                    8,
                    TreeNode(4)
                )
            )
        )
    )
    ...

File: contests/16_05_2020/3.go
package main

import (
	"fmt"
	"sync"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MinUint = 0
const MaxUint = ^uint(MinUint)
const MaxInt = int(MaxUint >> 1)
const MinInt = -MaxInt - 1

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MinUint = 0
const MaxUint = ^uint(MinUint)
const MaxInt = int(MaxUint >> 1)
const MinInt = -MaxInt - 1

func goodNodes(root *TreeNode) int {
	var ans int
	var lst []int

	wg.Add(1)
	preorder(root, lst, &ans)

	ans++
	return ans
}

var mux sync.Mutex
var wg sync.WaitGroup

func maxTwo(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func max(lst *[]int) (maxInLst int) {
	maxInLst = MinInt
	if len(*lst) == 0 {
		return
	}
	for _, v := range *lst {
		maxInLst = maxTwo(maxInLst, v)
	}
	return
}

func preorder(t *TreeNode, lst []int, ans *int) {
	defer wg.Done()
	if t == nil {
		return
	}
	maxInLst := max(&lst)
	if len(lst) > 0 && t.Val >= maxInLst {
		// deadlock / race-condition otherwise
		mux.Lock()
		defer mux.Unlock()
		*ans++
	}
	lst = append(lst, t.Val)
	wg.Add(1)
	go preorder(t.Left, lst, ans)
	// preorder(t.Left, lst, ans)
	wg.Add(1)
	go preorder(t.Right, lst, ans)
	// preorder(t.Right, lst, ans)
}

func main() {
	t := TreeNode{
		Val: 2,
		Right: &TreeNode{
			Val: 4,
			Left: &TreeNode{
				Val: 10,
			},
			Right: &TreeNode{
				Val: 8,
				Left: &TreeNode{
					Val: 4,
				},
			},
		},
	}
	ans := goodNodes(&t)
	fmt.Println(ans)
}
